faq.md

```diff
@@ -499,6 +499,81 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
 profile or release mode using the menu entries **Run > Flutter Run in Profile Mode**
 or **Release Mode**.
 
+### What programming paradigm does Flutter's framework use?
+
+Flutter is a multiparadigm programming environment. Many programming techniques
+developed over the past few decades are used where they are appropriate, to solve
+problems in a clean way. In no particular order:
+
+* Composition: The primary paradigm used by Flutter is that of using small objects with narrow
+scopes of behavior, composed together to obtain more complicated effects. Most widgets
+in the Flutter widget library are built in this way.
+
+* Functional programming: Entire applications can be built with only
+[StatelessWidget](https://docs.flutter.io/flutter/widgets/StatelessWidget-class.html)s,
+which are essentially functions that describe how arguments map to other functions,
+bottoming out in primitives that compute layouts or paint graphics.
+Additionally, heavy use is made of immutable data structures, including the entire
+[Widget](https://docs.flutter.io/flutter/widgets/Widget-class.html)
+class hierarchy as well as numerous supporting classes such as
+[Rect](https://docs.flutter.io/flutter/dart-ui/Rect-class.html) and
+[TextStyle](https://docs.flutter.io/flutter/painting/TextStyle-class.html).
+On a smaller scale, Dart's
+[Iterable](https://docs.flutter.io/flutter/dart-core/Iterable-class.html) API,
+which makes heavy use of the functional style (map, reduce, where, etc), is
+frequently used to process lists of values in the framework.
+
+* Event-driven programming: User interactions are represented by event objects that are
+dispatched to callbacks registered with event handlers. Screen updates are triggered
+by a similar callback mechanism. The [Listenable](https://docs.flutter.io/flutter/foundation/Listenable-class.html)
```

Comment:
I would be much happier if it used some kind of Observable interface, so the callbacks did not have to be called explicitly. This introduces a dependency that can get in the way of many optimizations.

Reply:
I'm not familiar with the Listenable interface, so let me give the underlying issue I'm concerned about. Separation of concerns wants to separate the instigator of an action from the handler of that action. A procedure does this for the handler (i.e., a function foo() does not know or care who calls it). While an event does it for the instigator (an event doesn't know or care who handles it). In reactive programming, you care about the latter. This can be done with "event-driven programming" but this is nothing new. It also requires an event to be created for every possible thing that might need to be handled. I was using the term "Observable interface" in the general sense (not Dart specific) to mean an interface that can be easily added to any class to allow it to be observed (e.g., one or more handlers subscribed to it). Does Listenable do this?

I was confused because your language seems to be describing Listenable as some kind of callback mechanism, which is not an Observable interface. In fact, I would not want for us to brag that using callbacks is "event-driven programming". In a callback, there is little separation of concerns.

This is especially confusing because in Flutter, a stateful widget has to explicitly call setState. For example, in the default Flutter app, inside class _MyHomePageState there is a comment that says: "This call to setState tells the Flutter framework that something has changed in this State". That is not event-driven programming (and is certainly not Reactive) because the state (the observable object) has to explicitly tell the framework that it changed). It is also not functional. Would the use of a Listenable interface fix this? I don't see one being used here.

Reply:
> This can be done with "event-driven programming" but this is nothing new.

That's rather the thesis of this entire FAQ entry. As the introduction says, "Many programming techniques developed over the past few decades are used in Flutter". It's not trying to say that we invented any of them.

> I was using the term "Observable interface" in the general sense (not Dart specific) to mean an interface that can be easily added to any class to allow it to be observed (e.g., one or more handlers subscribed to it). Does Listenable do this?

Yes. For example, Animation (and its subclasses such as AnimationController) implements Listenable, and when you want to be driven by an animation you subscribe to it. The Animation doesn't know anything about who is subscribed to it.

> a stateful widget has to explicitly call setState. [...] That is not event-driven programming

Agreed, setState is unrelated to this.